log/tokenizer_test.go

```diff
@@ -7,41 +7,64 @@ import (
 )
 
 func TestTokenizer(t *testing.T) {
-	tokens, err := tokenize("loki=something,s.e=2231,s=12,12=3,a=[1,2,3],b=[1],s=c")
-	assert.Equal(t, []token{
+	tests := []struct {
+		name         string
+		input        string
+		wantTokens   []token
+		wantErrorMsg string
+	}{
 		{
-			key:   "loki",
-			value: "something",
+			name:  "good keys should tokenize without error",
+			input: "loki=something,s.e=2231,s=12,12=3,a=[1,2,3],b=[1],s=c",
+			wantTokens: []token{
+				{key: "loki", value: "something"},
+				{key: "s.e", value: "2231"},
+				{key: "s", value: "12"},
+				{key: "12", value: "3"},
+				{key: "a", value: "1,2,3", inside: '['},
+				{key: "b", value: "1", inside: '['},
+				{key: "s", value: "c"},
+			},
 		},
 		{
-			key:   "s.e",
-			value: "2231",
+			name:         "empty value should error",
+			input:        "empty=",
+			wantErrorMsg: "key `empty=` with no value",
 		},
 		{
-			key:   "s",
-			value: "12",
+			name:         "unclosed array should error",
+			input:        "foo=[1,2,3",
+			wantErrorMsg: "array value for key `foo` didn't end",
 		},
 		{
-			key:   "12",
-			value: "3",
+			name:         "bad characters after array should error",
+			input:        "foo=[1,2,3]bar",
+			wantErrorMsg: "there was no ',' after an array with key 'foo'",
 		},
 		{
-			key:    "a",
-			value:  "1,2,3",
-			inside: '[',
+			name:         "empty key should error",
+			input:        ",foo=bar",
+			wantErrorMsg: "key `` with no value",
 		},
-		{
-			key:    "b",
-			value:  "1",
-			inside: '[',
-		},
-		{
-			key:   "s",
-			value: "c",
-		},
-	}, tokens)
-	assert.NoError(t, err)
+	}
+
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
 
-	_, err = tokenize("empty=")
-	assert.EqualError(t, err, "key `empty=` with no value")
+			gotTokens, err := tokenize(test.input)
+			if test.wantTokens != nil {
+				if err != nil {
+					t.Fatal("didn't expect error")
+				}
+				assert.Equal(t, test.wantTokens, gotTokens)
+			} else {
+				if err == nil {
+					t.Fatal("expected an error but got none")
+				}
+				assert.EqualError(t, err, test.wantErrorMsg)
+			}
```

Comment:
I think we could simplify this logic if we divide the test in two different functions TestTokenizer and TestTokenizerInvalid / Error.

In this way, we can simplify this `if-else` logic and remove the `name` field, because the name will be pretty obvious from the function name and from the expected error message. @mstoykov WDYT?

Reply:
Sure happy to split it up into two tests for clarity. I like having the name field because it sets the name of the sub-test with `t.Run(test.name, ...)`. When a particular test fails you can find the test case data quickly because of the test name is even more specific (e.g `TestTokenInvalid/empty_value_should_error` failed quickly gets you the test case and the particular case that failed).

Reply:
Broken up into two tests. Happy to drop the names too if you want. Limme know :)

Reply:
The assert library will print the expected error m,essage so it should be easy to catch. In any case, having it it's fine so we can keep it :+1: